ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -7493,9 +7507,17 @@ protected Operator genFileSinkPlan(String dest, QB qb, Operator input)
 
       // Add NOT NULL constraint check
       input = genConstraintsPlan(dest, qb, input);
+      if (destinationTable.getStorageHandler() != null && destinationTable.getStorageHandler().alwaysUnpartitioned()) {
+        partSpec = qbm.getPartSpecForAlias(dest);
+      }
 
       if (!qb.getIsQuery()) {
-        input = genConversionSelectOperator(dest, qb, input, destinationTable.getDeserializer(), dpCtx, null, destinationTable);
+        if (!updating(dest) && !deleting(dest)) {
+          input = genConversionSelectOperatorByAddPartition(dest, qb, input,
+                  destinationTable.getDeserializer(), destinationTable, partSpec);
+        }
+        input = genConversionSelectOperator(dest, qb, input, destinationTable.getDeserializer(), dpCtx, null,
+                destinationTable);
```

Comment:
Two conversion operators are generated if this is not a delete nor an update statement. Is it necessary?

Reply:
@kasakrisz The idea is to isolate the addition of partition column into the operator with other set of conditions.